// amazeui-touch/lib/Button.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _InternalPropTypes = require('./InternalPropTypes');

var _ClassNameMixin = require('./mixins/ClassNameMixin');

var _ClassNameMixin2 = _interopRequireDefault(_ClassNameMixin);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var Button = (0, _createReactClass2.default)({
  displayName: 'Button',
  mixins: [_ClassNameMixin2.default],

  propTypes: {
    classPrefix: _propTypes2.default.string.isRequired,
    component: _InternalPropTypes.component,
    href: _propTypes2.default.string,
    target: _propTypes2.default.string,
    amStyle: _propTypes2.default.string,
    amSize: _propTypes2.default.string,
    hollow: _propTypes2.default.bool,
    block: _propTypes2.default.bool,
    active: _propTypes2.default.bool,
    disabled: _propTypes2.default.bool
    // radius: PropTypes.bool,
    // rounded: PropTypes.bool,
  },

  getDefaultProps: function getDefaultProps() {
    return {
      classPrefix: 'btn'
    };
  },
  removeUnknownProp: function removeUnknownProp(props) {
    delete props.classPrefix;
    delete props.amStyle;
    delete props.amSize;
    delete props.hollow;
    delete props.block;
    delete props.active;

    return props;
  },
  renderAnchor: function renderAnchor(classes) {
    var _props = this.props,
        href = _props.href,
        Component = _props.component,
        children = _props.children,
        props = _objectWithoutProperties(_props, ['href', 'component', 'children']);

    Component = Component || 'a';

    href = href || '#';

    return _react2.default.createElement(
      Component,
      _extends({}, this.removeUnknownProp(props), {
        href: href,
        className: classes,
        role: 'button'
      }),
      children
    );
  },
  renderButton: function renderButton(classes) {
    var _props2 = this.props,
        Component = _props2.component,
        children = _props2.children,
        props = _objectWithoutProperties(_props2, ['component', 'children']);

    Component = Component || 'button';

    return _react2.default.createElement(
      Component,
      _extends({}, this.removeUnknownProp(props), {
        className: classes
      }),
      children
    );
  },
  render: function render() {
    var classSet = this.getClassSet();
    var _props3 = this.props,
        href = _props3.href,
        target = _props3.target,
        block = _props3.block,
        className = _props3.className;

    var renderType = href || target ? 'renderAnchor' : 'renderButton';

    // block button
    classSet[this.prefixClass('block')] = block;

    return this[renderType]((0, _classnames2.default)(classSet, className));
  }
});

exports.default = Button;
module.exports = exports['default'];
