// amazeui-touch/lib/Card.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _ClassNameMixin = require('./mixins/ClassNameMixin');

var _ClassNameMixin2 = _interopRequireDefault(_ClassNameMixin);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var Card = (0, _createReactClass2.default)({
  displayName: 'Card',
  mixins: [_ClassNameMixin2.default],

  propTypes: {
    classPrefix: _propTypes2.default.string.isRequired,
    title: _propTypes2.default.string,
    header: _propTypes2.default.node,
    footer: _propTypes2.default.node
  },

  getDefaultProps: function getDefaultProps() {
    return {
      classPrefix: 'card'
    };
  },
  renderItem: function renderItem(element, role) {
    if (!element) {
      return null;
    }

    return element.type && element.type === Card.Child ? element : _react2.default.createElement(
      Card.Child,
      { role: role },
      element
    );
  },
  renderTitle: function renderTitle(title) {
    return _react2.default.createElement(
      'h2',
      { className: this.prefixClass('title') },
      title
    );
  },
  render: function render() {
    var classSet = this.getClassSet();

    var _props = this.props,
        children = _props.children,
        className = _props.className,
        title = _props.title,
        header = _props.header,
        footer = _props.footer,
        props = _objectWithoutProperties(_props, ['children', 'className', 'title', 'header', 'footer']);

    delete props.classPrefix;

    return _react2.default.createElement(
      'div',
      _extends({}, props, {
        className: (0, _classnames2.default)(classSet, className)
      }),
      title ? this.renderItem(this.renderTitle(title)) : this.renderItem(header),
      _react2.default.createElement(
        'div',
        { className: this.prefixClass('body') },
        children
      ),
      this.renderItem(footer, 'footer')
    );
  }
});

var CardChild = (0, _createReactClass2.default)({
  displayName: 'CardChild',
  mixins: [_ClassNameMixin2.default],

  propTypes: {
    classPrefix: _propTypes2.default.string.isRequired,
    role: _propTypes2.default.oneOf(['header', 'footer']),
    cover: _propTypes2.default.string
  },

  getDefaultProps: function getDefaultProps() {
    return {
      classPrefix: 'card',
      role: 'header'
    };
  },
  render: function render() {
    var _classSet;

    var _props2 = this.props,
        role = _props2.role,
        className = _props2.className,
        cover = _props2.cover,
        props = _objectWithoutProperties(_props2, ['role', 'className', 'cover']);

    var classSet = (_classSet = {}, _defineProperty(_classSet, this.prefixClass(role), true), _defineProperty(_classSet, this.prefixClass('cover'), cover), _classSet);
    var style = null;

    if (cover) {
      style = {
        backgroundImage: 'url(' + cover + ')'
      };
    }

    delete props.classPrefix;

    return _react2.default.createElement(
      'div',
      _extends({}, props, {
        className: (0, _classnames2.default)(className, classSet),
        role: 'card-' + role,
        style: style
      }),
      this.props.children
    );
  }
});

Card.Child = CardChild;

exports.default = Card;
module.exports = exports['default'];
