// amazeui-touch/lib/Field.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _ClassNameMixin = require('./mixins/ClassNameMixin');

var _ClassNameMixin2 = _interopRequireDefault(_ClassNameMixin);

var _Button = require('./Button');

var _Button2 = _interopRequireDefault(_Button);

var _Icon = require('./Icon');

var _Icon2 = _interopRequireDefault(_Icon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var Field = (0, _createReactClass2.default)({
  displayName: 'Field',
  mixins: [_ClassNameMixin2.default],

  propTypes: {
    classPrefix: _propTypes2.default.string.isRequired,
    type: _propTypes2.default.string,
    label: _propTypes2.default.node,
    btnBefore: _propTypes2.default.node,
    btnAfter: _propTypes2.default.node,
    labelBefore: _propTypes2.default.node,
    labelAfter: _propTypes2.default.node,
    containerClassName: _propTypes2.default.string
  },

  getDefaultProps: function getDefaultProps() {
    return {
      classPrefix: 'field',
      type: 'text'
    };
  },
  getFieldDOMNode: function getFieldDOMNode() {
    return this.refs.field;
  },
  getValue: function getValue() {
    if (this.props.type === 'select' && this.props.multiple) {
      return this.getSelectedOptions();
    } else {
      return this.getFieldDOMNode().value;
    }
  },
  getChecked: function getChecked() {
    return this.getFieldDOMNode().checked;
  },
  getSelectedOptions: function getSelectedOptions() {
    var values = [];
    // see http://www.w3schools.com/jsref/coll_select_options.asp
    var options = this.getFieldDOMNode().options;

    Array.from(options).forEach(function (option) {
      if (option.selected) {
        var value = option.getAttribute('value') || option.innerHtml;

        values.push(value);
      }
    });

    return values;
  },
  isCheckboxOrRadio: function isCheckboxOrRadio() {
    return this.props.type === 'radio' || this.props.type === 'checkbox';
  },
  isFile: function isFile() {
    return this.props.type === 'file';
  },


  // convert `value`/`defaultValue` to `checked`/`defaultChecked` when `type` is `radio`/checkbox``
  convertValueToChecked: function convertValueToChecked() {
    var _this = this;

    var checkedProps = {};

    if (this.isCheckboxOrRadio()) {
      var propsMap = {
        checked: 'value',
        defaultChecked: 'defaultValue'
      };

      Object.keys(propsMap).forEach(function (checked) {
        var value = propsMap[checked];

        if (!_this.props[checked] && _this.props[value]) {
          checkedProps[checked] = value;
        }
      });
    }

    return checkedProps;
  },
  renderField: function renderField() {
    var field = null;
    var fieldClassName = this.isCheckboxOrRadio() || this.isFile() ? '' : this.getClassSet();
    var classes = (0, _classnames2.default)(this.props.className, fieldClassName);
    var commonProps = {
      ref: 'field',
      key: 'formField',
      className: classes
    };
    var assignedProps = _extends({}, this.props, commonProps);

    delete assignedProps.classPrefix;
    delete assignedProps.containerClassName;
    delete assignedProps.label;
    delete assignedProps.btnBefore;
    delete assignedProps.btnAfter;
    delete assignedProps.labelBefore;
    delete assignedProps.labelAfter;

    switch (this.props.type) {
      case 'select':
        field = _react2.default.createElement(
          'select',
          assignedProps,
          this.props.children
        );
        break;
      case 'textarea':
        field = _react2.default.createElement('textarea', assignedProps);
        break;
      case 'submit':
      case 'reset':
        var _props = this.props,
            classPrefix = _props.classPrefix,
            others = _objectWithoutProperties(_props, ['classPrefix']);

        field = _react2.default.createElement(_Button2.default, _extends({}, commonProps, {
          className: null
        }, others, {
          component: 'input'
        }));
        break;
      default:
        field = _react2.default.createElement('input', _extends({}, assignedProps, this.convertValueToChecked()));
    }

    return field;
  },
  renderContainer: function renderContainer(children) {
    var _props2 = this.props,
        id = _props2.id,
        label = _props2.label,
        containerClassName = _props2.containerClassName;

    return label ? _react2.default.createElement(
      'label',
      {
        htmlFor: id,
        className: (0, _classnames2.default)(this.prefixClass('container'), containerClassName),
        key: 'label'
      },
      _react2.default.createElement(
        'span',
        { className: this.prefixClass('label') },
        label
      ),
      children,
      this.isCheckboxOrRadio() ? _react2.default.createElement(_Icon2.default, {
        className: this.prefixClass('icon'),
        name: 'check'
      }) : null
    ) : children;
  },
  renderFieldGroup: function renderFieldGroup(children) {
    var _this2 = this;

    var groupPrefix = this.setClassNS('field-group');
    var labelClassName = groupPrefix + '-label';
    var _props3 = this.props,
        labelBefore = _props3.labelBefore,
        labelAfter = _props3.labelAfter,
        btnBefore = _props3.btnBefore,
        btnAfter = _props3.btnAfter,
        containerClassName = _props3.containerClassName;

    var renderFiledLabel = function renderFiledLabel(type) {
      return _this2.props[type] ? _react2.default.createElement(
        'span',
        {
          className: labelClassName,
          key: type
        },
        _this2.props[type]
      ) : null;
    };

    return labelBefore || labelAfter || btnBefore || btnAfter ? _react2.default.createElement(
      'div',
      {
        className: (0, _classnames2.default)(groupPrefix, containerClassName),
        key: 'fieldGroup'
      },
      renderFiledLabel('labelBefore'),
      btnBefore,
      children,
      renderFiledLabel('labelAfter'),
      btnAfter
    ) : children;
  },
  render: function render() {
    var field = this.renderField();

    if (this.props.label) {
      return this.renderContainer(field);
    }

    return this.renderFieldGroup(field);
  }
});

exports.default = Field;
module.exports = exports['default'];
