// amazeui-touch/lib/Grid.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _ClassNameMixin = require('./mixins/ClassNameMixin');

var _ClassNameMixin2 = _interopRequireDefault(_ClassNameMixin);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var Grid = (0, _createReactClass2.default)({
  displayName: 'Grid',
  mixins: [_ClassNameMixin2.default],

  propTypes: {
    classPrefix: _propTypes2.default.string.isRequired,
    component: _propTypes2.default.node.isRequired,
    collapse: _propTypes2.default.bool,
    avg: _propTypes2.default.number,
    align: _propTypes2.default.oneOf(['right', 'center', 'between', 'around']),
    wrap: _propTypes2.default.oneOf(['wrap', 'wrap-reverse']),
    bordered: _propTypes2.default.bool
  },

  getDefaultProps: function getDefaultProps() {
    return {
      classPrefix: 'g',
      component: 'div'
    };
  },


  render: function render() {
    var classSet = this.getClassSet();

    var _props = this.props,
        Component = _props.component,
        className = _props.className,
        collapse = _props.collapse,
        bordered = _props.bordered,
        avg = _props.avg,
        align = _props.align,
        wrap = _props.wrap,
        props = _objectWithoutProperties(_props, ['component', 'className', 'collapse', 'bordered', 'avg', 'align', 'wrap']);

    delete props.classPrefix;

    // .g-collapse
    classSet[this.prefixClass('collapse')] = collapse;

    // .g-bordered
    classSet[this.prefixClass('bordered')] = bordered;

    // .g-wrap
    classSet[this.prefixClass(wrap)] = wrap;

    if (avg) {
      classSet[this.prefixClass('avg-' + avg)] = true;
    }

    if (align) {
      classSet[this.prefixClass(align)] = true;
    }

    return _react2.default.createElement(
      Component,
      _extends({}, props, {
        className: (0, _classnames2.default)(className, classSet)
      }),
      this.props.children
    );
  }
});

exports.default = Grid;
module.exports = exports['default'];
