// amazeui-touch/lib/Loader.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _ClassNameMixin = require('./mixins/ClassNameMixin');

var _ClassNameMixin2 = _interopRequireDefault(_ClassNameMixin);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var Loader = (0, _createReactClass2.default)({
  displayName: 'Loader',
  mixins: [_ClassNameMixin2.default],

  propTypes: {
    classPrefix: _propTypes2.default.string,
    component: _propTypes2.default.node,
    amStyle: _propTypes2.default.string,
    rounded: _propTypes2.default.bool
  },

  getDefaultProps: function getDefaultProps() {
    return {
      classPrefix: 'loader',
      component: 'div'
    };
  },
  render: function render() {
    var classSet = this.getClassSet();

    var _props = this.props,
        className = _props.className,
        Component = _props.component,
        props = _objectWithoutProperties(_props, ['className', 'component']);

    delete props.classPrefix;
    delete props.amStyle;
    delete props.rounded;

    return _react2.default.createElement(
      Component,
      _extends({}, props, {
        className: (0, _classnames2.default)(classSet, className)
      }),
      _react2.default.createElement('div', { className: this.prefixClass('bounce1') }),
      _react2.default.createElement('div', { className: this.prefixClass('bounce2') }),
      _react2.default.createElement('div', { className: this.prefixClass('bounce3') })
    );
  }
});

exports.default = Loader;
module.exports = exports['default'];
