// amazeui-touch/lib/OffCanvasTrigger.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _reactDom = require('react-dom');

var _ClassNameMixin = require('./mixins/ClassNameMixin');

var _ClassNameMixin2 = _interopRequireDefault(_ClassNameMixin);

var _OverlayMixin = require('./mixins/OverlayMixin');

var _OverlayMixin2 = _interopRequireDefault(_OverlayMixin);

var _CSSCore = require('./utils/CSSCore');

var _CSSCore2 = _interopRequireDefault(_CSSCore);

var _TransitionEvents = require('./utils/TransitionEvents');

var _TransitionEvents2 = _interopRequireDefault(_TransitionEvents);

var _createChainedFunction = require('./utils/createChainedFunction');

var _createChainedFunction2 = _interopRequireDefault(_createChainedFunction);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var OffCanvasTrigger = (0, _createReactClass2.default)({
  displayName: 'OffCanvasTrigger',
  mixins: [_OverlayMixin2.default, _ClassNameMixin2.default],

  propTypes: {
    defaultOffCanvasActive: _propTypes2.default.bool,
    placement: _propTypes2.default.oneOf(['left', 'right']),
    animation: _propTypes2.default.oneOf(['slide', 'push']),
    offCanvas: _propTypes2.default.node.isRequired,
    pageContainer: _propTypes2.default.node,
    onOpen: _propTypes2.default.func,
    onClosed: _propTypes2.default.func
  },

  getDefaultProps: function getDefaultProps() {
    return {
      placement: 'left',
      animation: 'slide',
      onOpen: function onOpen() {},
      onClosed: function onClosed() {}
    };
  },
  getInitialState: function getInitialState() {
    return {
      offCanvasActive: this.props.defaultOffCanvasActive == null ? false : this.props.defaultOffCanvasActive,
      isClosing: false
    };
  },
  componentDidMount: function componentDidMount() {
    this.setPageContainer();
  },
  open: function open() {
    if (this.state.offCanvasActive) {
      return;
    }

    this.setState({
      offCanvasActive: true,
      isClosing: false
    }, function () {
      this.props.onOpen();
    });

    if (this.isPush()) {
      _CSSCore2.default.addClass(this.getContainerDOMNode(), this.getWithClassName());
    }
  },
  close: function close() {
    if (!this.state.offCanvasActive || this.state.isClosing) {
      return;
    }

    this.setState({
      isClosing: true
    });

    if (this.isPush()) {
      var container = this.getContainerDOMNode();
      _CSSCore2.default.removeClass(container, this.getWithClassName());
      _CSSCore2.default.addClass(container, this.getClosingClassName());
    }

    // 暂时取消动画
    this.handleClosed();
  },
  handleClosed: function handleClosed() {
    this.setState({
      offCanvasActive: false,
      isClosing: false
    });

    this.props.onClosed();

    if (this.isPush()) {
      _CSSCore2.default.removeClass(this.getContainerDOMNode(), this.getClosingClassName());
    }
  },
  toggle: function toggle() {
    this.state.offCanvasActive ? this.close() : this.open();
  },
  isPush: function isPush() {
    return this.props.animation === 'push';
  },
  getPageContainer: function getPageContainer() {
    var pageContainer = this.props.pageContainer;


    return typeof pageContainer === 'string' ? document.querySelector(pageContainer) : (0, _reactDom.findDOMNode)(pageContainer);
  },
  setPageContainer: function setPageContainer() {
    var pageContainer = this.getPageContainer();

    if (pageContainer && this.isPush()) {
      _CSSCore2.default.addClass(pageContainer, this.setClassNS('offcanvas-push-target'));
    }
  },
  getWithClassName: function getWithClassName() {
    return 'with-offcanvas-' + this.props.placement;
  },
  getClosingClassName: function getClosingClassName() {
    return 'with-offcanvas-closing';
  },


  // used by Mixin
  renderOverlay: function renderOverlay() {
    if (!this.state.offCanvasActive) {
      return _react2.default.createElement('span', null);
    }

    var offCanvas = this.props.offCanvas;
    var isClosing = this.state.isClosing;

    /*
    if (isClosing) {
      let node = this.getOverlayDOMNode();
      if (node) {
        let closedHandler = (e) => {
          if (e && e.target !== node) {
            return;
          }
           TransitionEvents.off(node, closedHandler);
          this.handleClosed();
        };
         TransitionEvents.on(node, closedHandler);
      } else {
        this.handleClosed();
      }
    }
    */

    return (0, _react.cloneElement)(offCanvas, {
      placement: this.props.placement,
      animation: this.props.animation,
      isClosing: isClosing,
      onDismiss: this.close
    });
  },
  render: function render() {
    var child = _react2.default.Children.only(this.props.children);
    var props = {
      onClick: (0, _createChainedFunction2.default)(child.props.onClick, this.props.onClick)
    };

    props.onClick = (0, _createChainedFunction2.default)(this.toggle, props.onClick);

    return (0, _react.cloneElement)(child, props);
  }
});

exports.default = OffCanvasTrigger;
module.exports = exports['default'];
