// amazeui-touch/lib/Tabs.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _ClassNameMixin = require('./mixins/ClassNameMixin');

var _ClassNameMixin2 = _interopRequireDefault(_ClassNameMixin);

var _Button = require('./Button');

var _Button2 = _interopRequireDefault(_Button);

var _ButtonGroup = require('./ButtonGroup');

var _ButtonGroup2 = _interopRequireDefault(_ButtonGroup);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var Tabs = (0, _createReactClass2.default)({
  displayName: 'Tabs',
  mixins: [_ClassNameMixin2.default],

  propTypes: {
    classPrefix: _propTypes2.default.string,
    activeKey: _propTypes2.default.any,
    defaultActiveKey: _propTypes2.default.any,
    onAction: _propTypes2.default.func,
    inset: _propTypes2.default.bool
  },

  getDefaultProps: function getDefaultProps() {
    return {
      classPrefix: 'tabs'
    };
  },
  getInitialState: function getInitialState() {
    return {
      activeKey: this.getDefaultActiveKey(),
      previousActiveKey: null
    };
  },
  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    var nextActiveKey = nextProps.activeKey;

    // update controlled Tabs' state
    if (nextActiveKey != null && nextActiveKey !== this.props.activeKey) {
      this.setState({
        activeKey: nextActiveKey,
        previousActiveKey: this.props.activeKey
      });
    }
  },
  getDefaultActiveKey: function getDefaultActiveKey(children) {
    var defaultActiveKey = this.props.defaultActiveKey;

    if (defaultActiveKey != null) {
      return defaultActiveKey;
    }

    _react2.default.Children.forEach(children, function (child) {
      if (defaultActiveKey == null) {
        defaultActiveKey = child.props.eventKey;
      }
    });

    return defaultActiveKey != null ? defaultActiveKey : 0;
  },
  getActiveKey: function getActiveKey() {
    return this.props.activeKey != null ? this.props.activeKey : this.state.activeKey;
  },
  handleClick: function handleClick(key, disabled, e) {
    e.preventDefault();
    var activeKey = this.state.activeKey;

    if (disabled) {
      return;
    }

    if (this.props.onAction) {
      this.props.onAction(key);
    }

    // uncontrolled
    if (this.props.activeKey == null && activeKey !== key) {
      this.setState({
        activeKey: key,
        previousActiveKey: activeKey
      });
    }
  },
  renderNav: function renderNav() {
    var _this = this;

    var activeKey = this.getActiveKey();

    var navs = _react2.default.Children.map(this.props.children, function (child, index) {
      var _child$props = child.props,
          eventKey = _child$props.eventKey,
          disabled = _child$props.disabled,
          navSize = _child$props.navSize,
          navStyle = _child$props.navStyle;

      var key = index;

      eventKey = eventKey !== undefined ? eventKey : index;
      var active = eventKey === activeKey;

      return _react2.default.createElement(
        _Button2.default,
        {
          ref: 'tabNav' + key,
          key: key,
          onClick: _this.handleClick.bind(_this, key, disabled),
          active: active,
          disabled: disabled,
          className: active ? 'active' : null,
          amSize: navSize || 'sm',
          amStyle: navStyle || 'primary',
          hollow: true
        },
        child.props.title
      );
    });

    return _react2.default.createElement(
      _ButtonGroup2.default,
      {
        className: this.prefixClass('nav'),
        justify: true
      },
      navs
    );
  },
  renderTabPanels: function renderTabPanels() {
    var activeKey = this.getActiveKey();
    var panels = _react2.default.Children.map(this.props.children, function (child, index) {
      var _child$props2 = child.props,
          eventKey = _child$props2.eventKey,
          children = _child$props2.children,
          props = _objectWithoutProperties(_child$props2, ['eventKey', 'children']);

      if (eventKey === undefined) {
        eventKey = index;
      }

      return _react2.default.createElement(
        Tabs.Item,
        _extends({
          active: eventKey === activeKey,
          eventKey: eventKey,
          key: 'tabPanel' + index
        }, props),
        children
      );
    });

    return _react2.default.createElement(
      'div',
      {
        className: this.prefixClass('body')
      },
      panels
    );
  },
  render: function render() {
    var classSet = this.getClassSet();

    var _props = this.props,
        className = _props.className,
        props = _objectWithoutProperties(_props, ['className']);

    delete props.classPrefix;
    delete props.activeKey;
    delete props.defaultActiveKey;
    delete props.inset;
    delete props.onAction;

    return _react2.default.createElement(
      'div',
      _extends({}, props, {
        className: (0, _classnames2.default)(classSet, className)
      }),
      this.renderNav(),
      this.renderTabPanels()
    );
  }
});

var TabsItem = (0, _createReactClass2.default)({
  displayName: 'TabsItem',
  mixins: [_ClassNameMixin2.default],

  propTypes: {
    classPrefix: _propTypes2.default.string,
    title: _propTypes2.default.node,
    eventKey: _propTypes2.default.any,
    disabled: _propTypes2.default.bool,
    active: _propTypes2.default.bool,
    noPadded: _propTypes2.default.bool,
    navSize: _propTypes2.default.string,
    navStyle: _propTypes2.default.string
  },

  getDefaultProps: function getDefaultProps() {
    return {
      classPrefix: 'tab'
    };
  },
  render: function render() {
    var classSet = this.getClassSet(true);

    var _props2 = this.props,
        className = _props2.className,
        children = _props2.children,
        noPadded = _props2.noPadded,
        props = _objectWithoutProperties(_props2, ['className', 'children', 'noPadded']);

    var elementName = 'panel';

    delete props.classPrefix;
    delete props.eventKey;
    delete props.active;
    delete props.noPadded;
    delete props.navSize;
    delete props.navStyle;

    classSet[this.prefixClass(elementName)] = true;
    classSet[this.prefixClass(elementName + '-no-padded')] = noPadded;

    return _react2.default.createElement(
      'div',
      _extends({}, props, {
        className: (0, _classnames2.default)(classSet, className)
      }),
      children
    );
  }
});

Tabs.Item = TabsItem;

exports.default = Tabs;

// TODO: Nav 的可定制性，如允许传入 Router 的 Link 组件

module.exports = exports['default'];
